### PR TITLE
Vertex Modification

### PR DESCRIPTION
**Pull request by Kink3d**
Image

# Vertex Modification

## Purpose

Adds ability to modify absolute vertex position to all master nodes.

## Changes

- Add `ShaderStageCapability` to all material slots
- Add `IsCompatibleStageWith` method to determine if two slots have compatible stages
- Add `ShaderStageCapability` to all mesh data requirement interfaces (needed for master node where different stages have different requirements)
- Add `ShaderStageExtensions`
- Add `includeIntermediateSpaces` check to `FromNodes<T>` (needed to exclude intermediate space translations from vertex and surface description input generation)
- Update `GetEffectiveShaderStage` and `GetEffectiveShaderStageCapability`to handle backwards edge detection (?)
- Add `Position` slot and `IMayRequirePosition` interface to both master nodes
- Add stage capability check to `IMayRequirePosition` and `IMayRequireNormal` on both master nodes
- Refactor master node shader generation to use `ShaderStringBuilder`
- Add `ShaderStageCapability` to `CodeFunctionNode` `SlotAttribute`
- Refactor `SurfaceMaterialOptions` to use `ShaderStringBuilder`
- Refactor `GetPropertiesDeclaration` (simpler)
- Refactor `ShaderGenerator` methods to use `ShaderStringBuilder`
- Major refactor to `GenerateStandardTransforms`
   - Add vertex requirements
   - Add generation of translations for vertex description
   - Add generation of vertex description inputs
   - Move everything to `ShaderStringBuilder`
   - Refactor code order for readability
   - Move all transformation code into a single block in output shader
- Refactor `GenerateSpaceTranslations` for vertex description use
- Refactor `GenerateSpaceTranslations` to use `ShaderStringBuilder`
- Refactor `GetPreviewSubShader` to use `ShaderStringBuilder`
- Refactor preview shader template for output readability
- Rename some `ShaderGeneratorNames` for consistency
- Add `BlockSemicolon` scope to `ShaderStringBuilder`
- Add indent level control to `IncreaseIndent` and `DecreaseIndent` in `ShaderStringBuilder`
- Add stage compability check to `SearchWindowProvider`
- Add initial framework for `Wireframe` type preview rendering
- Adds `Position` slot to lightweight subshaders
- Adds `VertexPopulation` stage for modifying vertex position
- Add graph control of vertex stage transformations
- Adds vertex slot and requirement collection
- Adds surface requirements (removes intermediate spaces from surface description inputs)
- Refactor final shader build order for better output layout
- Refactor pass generation methods for readability
- Refactor subshaders to use `ShaderStringBuilder` for generation
- Update all templates to support new features
- Update all templates for better final shader layout and formatting
- Refactor `HDUnlitSubShader`
   - Move everything to `ShaderStringBuilder`
   - Refactor shader generation for output readability
   - Add `Alpha Threshold` slot and alpha clip functionality
   - Refactor C# layout to match LW subshaders
- Refactor `HDUnlitPassForward.template` to match changes to subshader
- Define various nodes as fragment stage only
   - Sample Texture 2D
   - Triplanar
   - Checkerboard
   - Create Normal
   - Ellipse
   - Rectangle
   - Rounded Rectangle
   - Polygon

## Risks

- Touches a tonne of general shader generation code
- Refactors lots of `ShaderGenerator` in to `ShaderStringBuilder`
- Lots of template modification
- Affects all `IMayRequireX` interfaces
- No subgraph support for vertex (disabled for now). Fix in later PR

### CODE REVIEW

**Comment by isurfraz** [comment]
would love to see this go in!

File: ScriptableRenderPipeline/LightweightPipeline/LWRP/Editor/ShaderGraph/lightweightPBRForwardPass.template [hunk 1]
```diff
@@ -1,33 +1,29 @@
...
+	// Material options generated by graph
+${Tags}
+${Blending}
+${Culling}
+${ZTest}
+${ZWrite}
 	HLSLPROGRAM
     // Required to compile gles 2.0 with standard srp library
     #pragma prefer_hlslcc gles
-	#pragma exclude_renderers d3d11_9x
```
Comment (phi-lira) [inline ScriptableRenderPipeline/LightweightPipeline/LWRP/Editor/ShaderGraph/lightweightPBRForwardPass.template]: Why is this being removed? We need it to build to UWP.

File: ScriptableRenderPipeline/LightweightPipeline/LWRP/Editor/ShaderGraph/lightweightPBRForwardPass.template [hunk 2]
```diff
@@ -1,33 +1,29 @@
...
-	#pragma multi_compile _ _SHADOWS_ENABLED
-    #pragma multi_compile _ _LOCAL_SHADOWS_ENABLED
-    #pragma multi_compile _ _SHADOWS_SOFT
+	#pragma multi_compile _ FOG_LINEAR FOG_EXP2
 
 	// -------------------------------------
 	// Unity defined keywords
 	#pragma multi_compile _ DIRLIGHTMAP_COMBINED
 	#pragma multi_compile _ LIGHTMAP_ON
-	#pragma multi_compile_fog
```
Comment (phi-lira) [inline ScriptableRenderPipeline/LightweightPipeline/LWRP/Editor/ShaderGraph/lightweightPBRForwardPass.template]: Also here

File: ScriptableRenderPipeline/LightweightPipeline/LWRP/Editor/ShaderGraph/lightweightPBRExtraPasses.template [hunk 3]
```diff
@@ -1,89 +1,143 @@
...
-	#pragma multi_compile_instancing
-
-	#pragma vertex ShadowPassVertex
-	#pragma fragment ShadowPassFragment
-
-	#include "LWRP/ShaderLibrary/InputSurfacePBR.hlsl"
-	#include "LWRP/ShaderLibrary/LightweightPassShadow.hlsl"
-
-	ENDHLSL
+    Tags{"LightMode" = "ShadowCaster"}
```
Comment (phi-lira) [inline ScriptableRenderPipeline/LightweightPipeline/LWRP/Editor/ShaderGraph/lightweightPBRExtraPasses.template]: This is overwritting the ShadowCaster pass with an old version of it.

File: ScriptableRenderPipeline/LightweightPipeline/LWRP/Editor/ShaderGraph/lightweightPBRForwardPass.template (hunk 1 above)
Comment (Kink3d) [reply ScriptableRenderPipeline/LightweightPipeline/LWRP/Editor/ShaderGraph/lightweightPBRForwardPass.template]: The branch was older than these changes. Got messed up somehow in the merge. All these are fixed now.

**Comment by Kink3d** [comment]
Compiles and loads previously serialised shaders (Party Preview) on 18.2 and trunk (Win10 and OSX). Merging.

**Pull request merged**